### PR TITLE
Migrate to OpenSSL3

### PR DESCRIPTION
**Pull request by huwcbjones**
This PR:
* Drops support for OpenSSL < 3.0
* Drops testing on 32-bit platforms
* Removes use of deprecated symbols
* Removes a lot of transparent shim methods

### CODE REVIEW

File: object.go [hunk 1]
```diff
@@ -14,11 +14,18 @@
...
-// #include "shim.h"
+// #include <openssl/objects.h>
 import "C"
+import "unsafe"
 
 // CreateObjectIdentifier creates ObjectIdentifier and returns NID for the created
 // ObjectIdentifier
 func CreateObjectIdentifier(oid string, shortName string, longName string) NID {
-	return NID(C.OBJ_create(C.CString(oid), C.CString(shortName), C.CString(longName)))
+	cOid := C.CString(oid)
```
Comment (w-miller) [inline object.go:24]: I'm guessing we don't run any of the tests in this repo under valgrind currently? If not, probably worth setting that up -- writing/using C bindings without any memory checking is a dangerous game...

**Review by w-miller — APPROVED** [review]
A few minor comments but generally LGTM from what I can tell 👍

File: object.go (hunk 1 above)
Comment (huwcbjones) [reply object.go:24]: No we don't, I manually went through to audit all the places we're calling CString to ensure we free them at a later date.

Move to a new issue?

File: provider.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		c.ctx = nil
+	}
+}
+
+// GetNonFIPSCtx gets a non-FIPS context
+func GetNonFIPSCtx(withLegacy bool) (*LibraryContext, error) {
+	if nonFIPSCtx == nil {
+		ctx := C.OSSL_LIB_CTX_new()
+		if ctx == nil {
+			return nil, fmt.Errorf("foo")
```
[the remark this answers is not in the document]
Comment (huwcbjones) [reply provider.go]: resolved in later commit

File: object.go (hunk 1 above)
Comment (w-miller) [reply object.go:24]: https://github.com/pexip/go-openssl/issues/6

File: provider.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import (
+	"fmt"
+	"runtime"
+	"sync"
+	"unsafe"
+)
+
+var (
+	nonFIPSCtx       *LibraryContext = nil
+	nonFIPSLegacyCtx *LibraryContext = nil
```
[the remark this answers is not in the document]
Comment (huwcbjones) [reply provider.go]: Have had a go, lmk what you think!

[the remark this answers is not in the document]
Comment (w-miller) [reply provider.go]: Perfect!

**Pull request merged**